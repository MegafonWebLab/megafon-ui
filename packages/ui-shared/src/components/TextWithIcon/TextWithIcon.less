@import '~@megafon/ui-core/styles/base.less';

@b: e('.mfui-text-with-icon');

@{b} {
    &__header {
        margin-bottom: 24px;

        @media @mobileSM {
            text-align: center;
        }
    }
}
