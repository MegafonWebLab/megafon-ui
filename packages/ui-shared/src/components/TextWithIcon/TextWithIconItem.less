@import '~@megafon/ui-core/styles/base.less';

@b: e('.mfui-text-with-icon-item');

@{b} {
    display: flex;
    align-items: center;

    @media @mobileSM {
        flex-direction: column;
    }

    &:not(:first-of-type) {
        margin-top: 16px;

        @media @mobileSM {
            margin-top: 24px;
        }
    }

    &__svg-icon {
        width: 40px;
        min-width: 40px;
        height: 40px;
        min-height: 40px;
        overflow: hidden;

        @media @mobileBU {
            margin-right: 16px;
        }
    }

    &__text {
        @media @mobileSM {
            margin-top: 12px;

            text-align: center;
        }
    }
}
