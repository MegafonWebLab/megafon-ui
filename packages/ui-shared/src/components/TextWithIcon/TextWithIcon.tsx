import * as React from 'react';
import { Header, Grid, GridColumn } from '@megafon/ui-core';
import { cnCreate, filterDataAttrs, IFilterDataAttrs } from '@megafon/ui-helpers';
import * as PropTypes from 'prop-types';
import { ITextWithIconItem } from './TextWithIconItem';
import './TextWithIcon.less';

export interface ITextWithIconProps extends IFilterDataAttrs {
    /** Заголовок */
    title?: string;
    /** Ссылка на корневой элемент */
    rootRef?: React.Ref<HTMLDivElement>;
    /** Дополнительный класс для корневого элемента */
    className?: string;
    /** Допустимый дочерний компонент */
    children: React.ReactElement<ITextWithIconItem>[] | React.ReactElement<ITextWithIconItem>;
}

const cn = cnCreate('mfui-text-with-icon');
const TextWithIcon: React.FC<ITextWithIconProps> = ({ title, rootRef, dataAttrs, className, children }) => (
    <div className={cn([className])} ref={rootRef} {...filterDataAttrs(dataAttrs)}>
        <Grid>
            <GridColumn {...{ mobile: '12', tablet: '7', desktop: '6', wide: '6' }}>
                {title && (
                    <Header className={cn('header')} as="h5">
                        {title}
                    </Header>
                )}
                {children}
            </GridColumn>
        </Grid>
    </div>
);

TextWithIcon.propTypes = {
    title: PropTypes.string,
    rootRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.oneOfType([PropTypes.shape({ current: PropTypes.elementType }), PropTypes.any]),
    ]),
    children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.element.isRequired), PropTypes.element.isRequired])
        .isRequired,
    dataAttrs: PropTypes.objectOf(PropTypes.string.isRequired),
    className: PropTypes.string,
};

export default TextWithIcon;
