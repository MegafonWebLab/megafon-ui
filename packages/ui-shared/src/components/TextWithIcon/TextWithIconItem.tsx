import * as React from 'react';
import { cnCreate, convert, filterDataAttrs, textConvertConfig } from '@megafon/ui-helpers';
import * as PropTypes from 'prop-types';
import './TextWithIconItem.less';

export interface ITextWithIconItem {
    /** Дополнительные классы для внутренних элементов */
    classes?: {
        icon?: string;
        text?: string;
    };
    /** Текст */
    text: string | string[];
    /** Иконка */
    icon: React.ReactNode;
    /** Ссылка на корневой элемент */
    rootRef?: React.Ref<HTMLDivElement>;
    /** Дополнительные data атрибуты к внутренним элементам */
    dataAttrs?: {
        root?: Record<string, string>;
    };
    /* @deprecated */
    /** Выравнивание по центру на мобильных */
    centeringOnMobile?: boolean;
    /** Дополнительный класс для корневого элемента */
    className?: string;
}

const cn = cnCreate('mfui-text-with-icon-item');
const TextWithIconItem: React.FC<ITextWithIconItem> = ({
    className,
    classes = {},
    text,
    icon,
    rootRef,
    dataAttrs,
    centeringOnMobile = true,
}) => {
    const renderText = React.useMemo(() => {
        if (Array.isArray(text)) {
            return text.map(item => convert(item, textConvertConfig));
        }

        return convert(text, textConvertConfig);
    }, [text]);

    return (
        <div
            className={cn({ 'centering-on-mobile': centeringOnMobile }, [className])}
            ref={rootRef}
            {...filterDataAttrs(dataAttrs?.root)}
        >
            <div className={cn('svg-icon', [classes.icon])}>{icon}</div>
            <div className={cn('text', [classes.text])}>{renderText}</div>
        </div>
    );
};

TextWithIconItem.propTypes = {
    text: PropTypes.oneOfType([PropTypes.string.isRequired, PropTypes.arrayOf(PropTypes.string.isRequired)]).isRequired,
    icon: PropTypes.node.isRequired,
    rootRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.oneOfType([PropTypes.shape({ current: PropTypes.elementType }), PropTypes.any]),
    ]),
    dataAttrs: PropTypes.shape({
        root: PropTypes.objectOf(PropTypes.string.isRequired),
    }),
    centeringOnMobile: PropTypes.bool,
    className: PropTypes.string,
};

export default TextWithIconItem;
