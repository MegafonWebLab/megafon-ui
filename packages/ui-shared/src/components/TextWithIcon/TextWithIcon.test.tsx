import * as React from 'react';
import Heart from '@megafon/ui-icons/basic-16-heart_16.svg';
import Star from '@megafon/ui-icons/basic-16-promo_16.svg';
import { shallow, mount } from 'enzyme';
import TextWithIcon from './TextWithIcon';
import TextWithIconItem from './TextWithIconItem';

const children = [
    <TextWithIconItem text="First item" icon={<Star style={{ fill: '#00B956' }} />} key={1} />,
    <TextWithIconItem text="Second item" icon={<Heart style={{ fill: '#00B956' }} />} key={2} />,
];

describe('<TextWithIcon />', () => {
    it('should render', () => {
        const wrapper = shallow(<TextWithIcon>{children}</TextWithIcon>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render with props', () => {
        const wrapper = shallow(
            <TextWithIcon title="Test title" className="test-class">
                {children}
            </TextWithIcon>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should return reference to root element', () => {
        const ref: React.RefObject<HTMLDivElement> = React.createRef();
        mount(<TextWithIcon rootRef={ref}>{children}</TextWithIcon>);

        expect(ref.current).not.toBeNull();
    });
});
