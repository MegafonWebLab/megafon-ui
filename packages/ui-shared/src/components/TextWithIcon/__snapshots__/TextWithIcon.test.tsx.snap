// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextWithIcon /> should render 1`] = `
<div
  className="mfui-text-with-icon"
>
  <Grid>
    <GridColumn
      desktop="6"
      mobile="12"
      tablet="7"
      wide="6"
    >
      <Header
        as="h5"
        className="mfui-text-with-icon__header"
      >
        Test title
      </Header>
      <TextWithIconItem
        icon={
          <test-file-stub
            style={
              Object {
                "fill": "#00B956",
              }
            }
          />
        }
        key="1"
        text="First item"
      />
      <TextWithIconItem
        icon={
          <test-file-stub
            style={
              Object {
                "fill": "#00B956",
              }
            }
          />
        }
        key="2"
        text="Second item"
      />
    </GridColumn>
  </Grid>
</div>
`;

exports[`<TextWithIcon /> should render with props 1`] = `
<div
  className="mfui-text-with-icon test-class"
>
  <Grid>
    <GridColumn
      desktop="6"
      mobile="12"
      tablet="7"
      wide="6"
    >
      <Header
        as="h5"
        className="mfui-text-with-icon__header"
        color="black"
        hAlign="inherit"
      >
        Test title
      </Header>
      <TextWithIconItem
        icon={
          <test-file-stub
            style={
              Object {
                "fill": "#00B956",
              }
            }
          />
        }
        key="1"
        text="First item"
      />
      <TextWithIconItem
        icon={
          <test-file-stub
            style={
              Object {
                "fill": "#00B956",
              }
            }
          />
        }
        key="2"
        text="Second item"
      />
    </GridColumn>
  </Grid>
</div>
`;
